### ISSUE TITLE
HTTP and HTTPS agent

### ISSUE THREAD

**Issue by normanrz** [bitinn/node-fetch#588]
I am trying to use keep-alive agent with node-fetch as described in #423, but I have a problem with HTTPS redirects. Initially, I call a HTTP resource which returns a redirect to a HTTPS resource. In the default `node-fetch` this is not a problem, because the agent is set based on the current request. When overriding the agent there is no way of having per-request agents.

I would propose an API that allows a function instead of a fixed agent:
```javascript
fetch(url, { agent: (protocol) => protocol === 'https' ? httpsAgent : httpAgent })
```

**Comment by bitinn** [comment]
Yes this is a known limit, but since `agent` is non-standard I doubt there will be much work on this, however, if you are lucky enough to be using nodejs 11.7.0+, globalAgent is now replaceable, meaning you can control this precisely:

- https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V11.md#2019-01-17-version-1170-current-bridgear
- https://github.com/nodejs/node/pull/25170

**Comment by normanrz** [comment]
Thanks for pointing to that change in node 11.

**Comment by bitinn** [comment]
this issue is addressed in v2.6.0, `options.agent` now accepts a function, see readme:

https://github.com/bitinn/node-fetch#custom-agent
